File: Android/app/src/main/java/com/auralis/crisisconnect/security/FirebaseRoleHelper.kt
package com.auralis.crisisconnect.security

import android.util.Log
import com.google.android.gms.tasks.Task
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseAuthException
import com.google.firebase.firestore.FirebaseFirestore
import java.util.Locale
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext

/**
 * Centralized helper that gates access to Rescue tooling by inspecting the authenticated user's
 * Firebase custom claims.
 */
object FirebaseRoleHelper {

    private const val TAG = "FirebaseRoleHelper"
    private val ALLOWED_ROLES = setOf("admin", "fieldteam")

    /**
     * Represents the result of checking a user's Rescue role from their Firebase ID token claims.
     */
    sealed class RescueRoleResult {
        data class Authorized(val role: String) : RescueRoleResult()
        object Unauthenticated : RescueRoleResult()
        object Unauthorized : RescueRoleResult()
        data class Failure(val exception: Throwable) : RescueRoleResult()
    }

    /**
     * Represents the result of rescue role authorization after re-checking Firebase ID token claims.
     */
    sealed class RescuePermissionResult {
        object Authorized : RescuePermissionResult()
        object Forbidden : RescuePermissionResult()
        object Unauthenticated : RescuePermissionResult()
        data class Failure(val exception: Throwable) : RescuePermissionResult()
    }

    suspend fun fetchRescueKeyFor(
        userId: String,
        auth: FirebaseAuth = FirebaseAuth.getInstance(),
        firestore: FirebaseFirestore = FirebaseFirestore.getInstance()
    ): ByteArray = withContext(Dispatchers.IO) {
        val caller = auth.currentUser?.takeUnless { it.isAnonymous }
            ?: throw IllegalStateException("User must be authenticated")
        val sanitizedUserId = userId.trim()
        require(sanitizedUserId.isNotEmpty()) { "User ID is required to fetch rescue key" }

        // Refresh the caller token to ensure Firestore security rules evaluate the latest claims.
        caller.getIdToken(true).awaitResult()
        val keyBytes = AesCipherHelper.loadAesKeyBase64(
            firestore = firestore,
            deviceId = sanitizedUserId,
            collectionPath = "users"
        )
        if (Log.isLoggable(TAG, Log.DEBUG)) {
            Log.d(TAG, "Fetched ${keyBytes.size * 8}-bit rescue key for $sanitizedUserId")
        }
        keyBytes
    }

    /**
     * Refreshes the caller's Firebase ID token and extracts the normalized `role` custom claim.
     * This executes on a background dispatcher to avoid touching the main thread.
     */
    suspend fun fetchRescueRole(
        auth: FirebaseAuth = FirebaseAuth.getInstance()
    ): RescueRoleResult = withContext(Dispatchers.IO) {
        val user = auth.currentUser?.takeUnless { it.isAnonymous }
            ?: return@withContext RescueRoleResult.Unauthenticated

        return@withContext try {
            val tokenResult = user.getIdToken(true).awaitResult()
            val rawRole = tokenResult.claims["role"] as? String
            val normalizedRole = rawRole?.lowercase(Locale.US)?.trim()

            if (normalizedRole != null && normalizedRole in ALLOWED_ROLES) {
                RescueRoleResult.Authorized(normalizedRole)
            } else {
                Log.i(TAG, "User lacks Rescue privileges. Claim=$rawRole")
                RescueRoleResult.Unauthorized
            }
        } catch (cancellation: CancellationException) {
            throw cancellation
        } catch (authException: FirebaseAuthException) {
            Log.w(TAG, "Firebase rejected role fetch", authException)
            RescueRoleResult.Failure(authException)
        } catch (throwable: Throwable) {
            Log.e(TAG, "Unexpected error while verifying role", throwable)
            RescueRoleResult.Failure(throwable)
        }
    }

    /**
     * Re-checks the current Firebase ID token for Rescue privilege.
     */
    suspend fun verifyRescuePermission(
        auth: FirebaseAuth = FirebaseAuth.getInstance()
    ): RescuePermissionResult = withContext(Dispatchers.IO) {
        val user = auth.currentUser?.takeUnless { it.isAnonymous }
            ?: return@withContext RescuePermissionResult.Unauthenticated
        when (val roleResult = fetchRescueRole(auth)) {
            is RescueRoleResult.Authorized -> RescuePermissionResult.Authorized
            RescueRoleResult.Unauthorized -> RescuePermissionResult.Forbidden
            RescueRoleResult.Unauthenticated -> RescuePermissionResult.Unauthenticated
            is RescueRoleResult.Failure -> RescuePermissionResult.Failure(roleResult.exception)
        }
    }
}

/**
 * Lightweight coroutine bridge so we can await Play Services tasks without extra dependencies.
 */
private suspend fun <T> Task<T>.awaitResult(): T = suspendCancellableCoroutine { continuation ->
    addOnCompleteListener { task ->
        if (task.isSuccessful) {
            val result = task.result
            if (result != null) {
                continuation.resume(result)
            } else {
                continuation.resumeWithException(
                    IllegalStateException("Firebase returned a null result")
                )
            }
        } else {
            val error = task.exception ?: IllegalStateException("Unknown Firebase task failure")
            continuation.resumeWithException(error)
        }
    }

    continuation.invokeOnCancellation {
        // Tasks API does not expose cancellation, but invokeOnCancellation allows cooperative cleanup.
    }
}

File: iOS/Crisis Connect/Services/Firebase/FirebaseRoleHelper.swift
//
//  FirebaseRoleHelper.swift
//  Crisis Connect
//
//  Created by Assistant on 28.12.2025
//

import Foundation
import FirebaseAuth

enum FirebaseRoleHelper {
    private static let rescueEndpoint = "https://verifyrescueaction-jxxgznalnq-uc.a.run.app"

    enum RescueRoleResult {
        case authorized(String)
        case unauthenticated
        case unauthorized
        case failure(Error)
    }

    enum RescuePermissionResult {
        case authorized
        case forbidden
        case unauthenticated
        case failure(Error)
    }

    static func fetchRescueRole(auth: Auth = Auth.auth()) async -> RescueRoleResult {
        guard let user = auth.currentUser, !user.isAnonymous else {
            return .unauthenticated
        }
        do {
            let token = try await tokenResult(for: user)
            let rawRole = token.claims["role"] as? String
            if let normalized = RescueRoleAccess.normalizedRole(rawRole) {
                return .authorized(normalized)
            }
            return .unauthorized
        } catch {
            return .failure(error)
        }
    }

    static func verifyRescuePermission(
        auth: Auth = Auth.auth(),
        endpoint: String? = nil
    ) async -> RescuePermissionResult {
        guard let user = auth.currentUser, !user.isAnonymous else {
            return .unauthenticated
        }
        do {
            let token = try await tokenResult(for: user)
            let idToken = token.token
            guard !idToken.isEmpty else {
                return .failure(NSError(
                    domain: "CrisisConnect.Firebase",
                    code: -3,
                    userInfo: [NSLocalizedDescriptionKey: "Firebase returned empty token."]
                ))
            }

            let resolvedEndpoint = endpoint ?? rescueEndpoint
            guard let url = URL(string: resolvedEndpoint) else {
                return .failure(NSError(
                    domain: "CrisisConnect.Firebase",
                    code: -4,
                    userInfo: [NSLocalizedDescriptionKey: "Invalid rescue endpoint URL."]
                ))
            }

            var request = URLRequest(url: url)
            request.httpMethod = "POST"
            request.timeoutInterval = 10
            request.setValue("Bearer \(idToken)", forHTTPHeaderField: "Authorization")

            let (_, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse else {
                return .failure(NSError(
                    domain: "CrisisConnect.Firebase",
                    code: -5,
                    userInfo: [NSLocalizedDescriptionKey: "Invalid rescue response."]
                ))
            }

            switch http.statusCode {
            case 200:
                return .authorized
            case 403:
                return .forbidden
            case 401:
                return .unauthenticated
            default:
                return .failure(NSError(
                    domain: "CrisisConnect.Firebase",
                    code: http.statusCode,
                    userInfo: [NSLocalizedDescriptionKey: "Unexpected response code \(http.statusCode)."]
                ))
            }
        } catch {
            return .failure(error)
        }
    }

    private static func tokenResult(for user: User) async throws -> AuthTokenResult {
        try await withCheckedThrowingContinuation { continuation in
            user.getIDTokenResult(forcingRefresh: true) { result, error in
                if let error = error {
                    continuation.resume(throwing: error)
                } else if let result = result {
                    continuation.resume(returning: result)
                } else {
                    continuation.resume(throwing: NSError(
                        domain: "CrisisConnect.Firebase",
                        code: -6,
                        userInfo: [NSLocalizedDescriptionKey: "Token result is nil."]
                    ))
                }
            }
        }
    }
}
